### PR TITLE
feat: doomsday tester

### PR DESCRIPTION
**Pull request by GalloDaSballo**
Doomsday Tester
- [x] Open DAO CDP
- [x] Allows Arbitrary opening and closing of "victim" cdp
- [ ] Cleanup
- [x] Do not perform the check during RM
- [x] Skip CDPs Closed for some other reason
- [ ] Run it for 2 hrs

### CODE REVIEW

File: packages/contracts/contracts/TestContracts/invariants/echidna/EchidnaDoomsdayTester.sol [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            assert(success);
+            (success, ) = actors[addresses[i]].proxy(
+                address(collateral),
+                abi.encodeWithSelector(CollateralTokenTester.deposit.selector, ""),
+                INITIAL_COLL_BALANCE
+            );
+            assert(success);
+        }
+
+        priceFeedMock.setPrice(1e8); /// TODO: Does this price make any sense?
```
Comment (dapp-whisperer) [inline packages/contracts/contracts/TestContracts/invariants/echidna/EchidnaDoomsdayTester.sol:45]: as a reviewer not sure as I'm not sure what the price is trying to achieve.

File: packages/contracts/contracts/TestContracts/invariants/echidna/EchidnaDoomsdayTester.sol [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        bytes32 newCdpId;
+        if (success) {
+            newCdpId = abi.decode(returnData, (bytes32));
+        }
+
+        return (success, newCdpId);
+    }
+
+    /// NOTE: Terrible code re-use but hopefully easy to understand
+    function victimOpenCdp(uint256 _col, uint256 _EBTCAmount) external {
```
Comment (dapp-whisperer) [inline packages/contracts/contracts/TestContracts/invariants/echidna/EchidnaDoomsdayTester.sol:102]: I think the macro is straitforward and fine here.

The feature is ultimately wrapping openCdp with tracking in our own cdp array, so a more general function might be:

```
openCdpWithTracking(coll, debt, trackingArray)
```

**Comment by dapp-whisperer** [comment]
# Doomsday WIthdraw
A minimal implementation of withdrawing ALL Cdps besides the final one, 

## Setup
For each actor (DAO and VICTIM)
- give them an initial ETH balance (make sure it succeeds!)
- take ETH and deposit some coll into stETH (no direct mint)

DAO then seeds 1 Cdp at CCR

### Actions
There are two fuzz actions:
- victim opens cdp
- victim closes all their cdps (DAO still has one open at CCR)

## Notes
The CDPs are always opened at the same ICR (CCR), the price never changes, and rebases never occur. Should we add more complexity here or to a variant

skipping withdraw loop in RM is good call, however as it stands we never enter RM.

**Comment by GalloDaSballo** [comment]
> ## Notes
> The CDPs are always opened at the same ICR (CCR), the price never changes, and rebases never occur. Should we add more complexity here or to a variant
> 
> skipping withdraw loop in RM is good call, however as it stands we never enter RM.

The `TargetFunctions` make it so that the price changes, RM is triggered, Redemptions are happening, etc..

Simply the Victim never does anything to their own CDPs

File: packages/contracts/medusa.json [hunk 3]
```diff
@@ -5,16 +5,16 @@
...
-			"EchidnaTester"
+			"EchidnaDoomsdayTester"
 		],
 		"targetContracts": [
-			"EchidnaTester"
+			"EchidnaDoomsdayTester"
 		],
 		"targetContractsBalances": [
-			"0x27b46536c66c8e3000000"
+			"0x4F68CA6D8CD91C6000000"
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply packages/contracts/medusa.json:17]: Yes we can, but at the same time the balances are not really used, as what's used is the Initial Coll Balance which is 1e21 vs the 1e24 initial ETH

Worth thinking about this or just setting this to an insanely high value

File: packages/contracts/contracts/TestContracts/invariants/echidna/EchidnaDoomsdayTester.sol (hunk 1 above)
Comment (GalloDaSballo) [reply packages/contracts/contracts/TestContracts/invariants/echidna/EchidnaDoomsdayTester.sol:45]: Please do review the amount opened on the CDP

I was having issues in minting 1e18+ eBTC

**Pull request merged**